laurat007/assignment2/Ancestors.java

```diff
@@ -0,0 +1,198 @@
+ *      1. add node in the left extremity
+ *      2. add node in the right extremity
+ *      3. add node in order to balance the tree with the possibility to skip 
+ *      insert locations
+ * 
+ *  The method solving the first task is a recursive one, similar to searching a
+ * node in a tree: when the given value is found, the method will return true, 
+ * but if the value was not found yet, we test if it was found in the current
+ * subtree (in either left or right subtree of the current root) - if this is 
+ * true we add the root to solution and return true, else return false.
+ * 
+ *  The method solving the second task uses the first method as follows:
+ * there are two values given; for each one it finds its ancestors and it 
+ * searches the first common value.
+ * 
+ */
+
+public class Ancestors {
+    static class BinaryTreeNode<T> {
+        BinaryTreeNode left, right;
+        T value;
+        
+        BinaryTreeNode(T value) {
+            this.value  = value;
+        }
+        
+        // add to left extremity
+        void addLeft(T newValue) {
+            BinaryTreeNode aux = this;
+            while(aux.left != null) {
+                aux = aux.left;
+            }
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            aux.left = newNode;
+        }
+        
+        // add to right extremity
+        void addRight(T newValue) {
+            BinaryTreeNode aux = this;
+            while(aux.right != null) {
+                aux = aux.right;
+            }
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            aux.right = newNode;
+        }
+        
+        // add to balance tree
+        // similar to bfs traversal
+        void addBalanced(T newValue, int skips) { 
+            BinaryTreeNode aux = this;
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            Queue<BinaryTreeNode> queue = new LinkedList();
+            queue.add(aux);
+            
+            while(true) {
+                BinaryTreeNode current = queue.poll();
+                if(current.left != null)
+                    queue.add(current.left);
+                else {
+                    if(skips != 0)
+                        skips--;
+                    else {
+                        current.left = newNode;
+                        break;
+                    }
+                }
+                if(current.right != null)
+                    queue.add(current.right);
+                else {
+                    if(skips != 0)
+                        skips--;
+                    else {
+                        current.right = newNode;
+                        break;
+                    }
+                }
+            }
+        }
+        
+        public String toString() {
+            String s = "";
+            BinaryTreeNode aux = this;
+            Queue<BinaryTreeNode> q = new LinkedList();
+            q.add(aux); 
```

Comment:
As above: you can remove the aux variable here.

Reply:
Done